### PR TITLE
Add plotting to common widget pose

### PR DESCRIPTION
**Pull request by AzulRadio**
Signed-off-by: youhy <haoyuan2019@outlook.com>

# New feature

As mentioned here https://github.com/gazebosim/gz-sim/pull/1571. This PR adds plotting to the common widget pose, then we can forward-port gz-sim citadel (3) to fortress (6) without a recession of features. 

## Summary

Should be exactly the same as before:

demo_plot

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: include/ignition/gui/qml/GzPlot.qml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+import QtQuick 2.9
+
+/**
+ * Item for a draggable plotting icon
+ */
```
Comment (jennuine) [inline include/ignition/gui/qml/GzPlot.qml:21]: Can you add a usage example?

File: include/ignition/gui/qml/GzPose.qml [hunk 2]
```diff
@@ -143,11 +153,37 @@ Item {
...
+        }
+        Component.onCompleted: {
+          xPlot.item.gzIconWidth = iconWidth
+          xPlot.item.gzIconHeight = iconHeight
+          xPlot.item.gzIconY = 10
+          xPlot.item.gzComponentInfo = "x"
+          xPlot.item.gzMimeData = { "text/plain" : (model === null) ? "" :
+              "Component," + model.entity + "," + model.typeId + "," +
+              model.dataType + "," + xPlot.item.gzComponentInfo + "," + model.shortName
+          }
```
Comment (jennuine) [inline include/ignition/gui/qml/GzPose.qml:175]: To me, this is very `gz-sim` specific (eg., `model.entity`, `model.typeId`, etc). How about we expose a variable for the mime data in `GzPose.qml` then pass the information from `gz-sim`?

What are your thoughts @chapulina ?

File: include/ignition/gui/qml/GzPose.qml [hunk 3]
```diff
@@ -124,6 +126,14 @@ Item {
     }
   }
 
+  /**
+   * Used to create a plotting icon
+   */
+  Component {
+    id: gzplot
```
Comment (jennuine) [inline include/ignition/gui/qml/GzPose.qml:133]:
```suggestion
    id: gzPlotPose
```

File: include/ignition/gui/qml/GzPlot.qml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+Item {
+  id: gzPlotIcon
+  property string gzComponentInfo
+  property var gzMimeData
+  property var gzDragTarget
```
Comment (jennuine) [inline include/ignition/gui/qml/GzPlot.qml:27]: can you add docs to what these variables are?

File: include/ignition/gui/qml/GzPlot.qml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  property int gzIconWidth : 20
+  property int gzIconHeight : 20
+  property int gzIconY : 10
```
Comment (jennuine) [inline include/ignition/gui/qml/GzPlot.qml:30]: nit:
```suggestion
  property int gzIconWidth: 20
  property int gzIconHeight: 20
  property int gzIconY: 10
```

**Review by jennuine — COMMENTED** [review]
There are some issues that need to be worked out.

This is how the plotting worked before:
before

This is after:
after

Currently, it's not able to plot any new data with the same component. For example, if plotting the `z` for the box and I want to add the `z` for the sphere nothing happens. I can only drag in a new component but the entity is incorrect (The box entity is `8` and the cylinder entity is `12` but in the legend it's showing `8`).

File: include/ignition/gui/qml/GzPose.qml [hunk 6]
```diff
@@ -124,6 +126,19 @@ Item {
...
 
+  /**
+   * Used to create a plotting icon
+   */
+  Component {
+    id: gzPlotPose
+    GzPlot {
+      gzMimeData : { "text/plain" : (model === null) ? "" :
+          "Component," + model.entity + "," + model.typeId + "," +
+          model.dataType + "," + gzComponentInfo + "," + model.shortName
```
Comment (jennuine) [inline include/ignition/gui/qml/GzPose.qml:137]: This is `gz-sim` specific. We should just expose `gzMimeData` and have `gz-sim` set this data

**Review by jennuine — COMMENTED** [review]
Nice work on the bug fixes, left some comments. Also, lets rename `GzPlot.qml` to `GzPlotIcon.qml` for clarity

File: include/ignition/gui/qml/GzPlot.qml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+ *   Rectangle {
+ *     Loader {
+ *       id: plotLoader
+ *       sourceComponent: gzplot
+ *     }
+ *     Component.onCompleted: {
+ *       plotLoader.item.gzMimeData = {"text/plain" : ""}
+ *     }
+ *   }
+ * }
```
Comment (jennuine) [inline include/ignition/gui/qml/GzPlot.qml:38]: This looks like it needs updated

**Review by jennuine — COMMENTED** [review]
After some investigation, I figured out what `mimeData` needs to have (see comments). Let's add docs to this to help future developers

File: include/ignition/gui/qml/GzPose.qml [hunk 8]
```diff
@@ -72,10 +78,14 @@ Item {
   // Expand/Collapse of this widget
   property bool expand: true
 
+  property var gzPlotMimeData
 
   /*** The following are private variables: ***/
   height: gzPoseContent.height
 
+  property int iconWidth: 20
+  property int iconHeight: 20
```
Comment (chapulina) [inline include/ignition/gui/qml/GzPose.qml:87]: How about keeping these inside `GzPlotIcon`? This way we ensure it's always consistent across widgets

File: include/ignition/gui/qml/GzPose.qml [hunk 9]
```diff
@@ -124,6 +142,16 @@ Item {
     }
   }
 
+  /**
+   * Used to create a plotting icon
+   */
+  Component {
+    id: gzPlotPose
+    GzPlot {
```
Comment (chapulina) [inline include/ignition/gui/qml/GzPose.qml:150]: Can we make the plot icon optional? This way the widget can be used in plugins that don't have plottable data, like `GridConfig`, `ViewAngle`, etc.

**Review by chapulina — COMMENTED** [review]
+1 to renaming to `GzPlotIcon`

**Comment by jennuine** [comment]
Thank you for iterating

4b89afe5cede3c2a1e9658d9ccd82dee2cb6135d I updated the optional technique to use `Loader.active` instead of a placeholder. I also set the default plot enabled to false and some other minor updates. Let me know if you're ok with this

As a quick test for the new optional plotting, you can use this PR https://github.com/gazebosim/gz-sim/pull/1607 and replace [this line in Pose3d.qml](https://github.com/gazebosim/gz-sim/blob/67dc38f445da5ade7af594a1a56a1f30c3b4499e/src/gui/plugins/component_inspector/Pose3d.qml#L120) with:
```javascript
gzPlotEnabled: {
  return entityType == "model"
}
```

Then the plotting icons will be hidden for links, etc:
Peek 2022-08-05 14-12

File: include/ignition/gui/qml/GzPlotIcon.qml [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+ *   }
+ * }
+ *
+ */
+Rectangle {
+  /**
+   * Used to transfer data, where data is separated by ','.
+   * For components, the syntax needs to be:
+   *   `{"text/plain" : "Component," + <entity> + "," + <typeId> + "," + <attribute> + "," + <typeName>}`.
+   * For fields: `{"text/plain" : <topic> + "," + <fieldPath>}`.
```
Comment (chapulina) [inline include/ignition/gui/qml/GzPlotIcon.qml:46]:
```suggestion
   * For transport topics: `{"text/plain" : <topic> + "," + <fieldPath>}`.
```

**Review by chapulina — APPROVED** [review]
Works for me!

**Pull request merged**